sig-scheduler-plugins/pkg/controllers/podgroup_controller.go

```diff
@@ -0,0 +1,473 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	schedv1alpha1 "sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+	fluenceLabels "sigs.k8s.io/scheduler-plugins/pkg/fluence/labels"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
+)
+
+// PodGroupReconciler reconciles a PodGroup object
+type PodGroupReconciler struct {
+	log      logr.Logger
+	recorder record.EventRecorder
+
+	client.Client
+	Scheme  *runtime.Scheme
+	Workers int
+}
+
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=scheduling.x-k8s.io,resources=podgroups/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Note that we currently don't do deletion based on owner references, but that
+// would be ideal (I could not get it to work)
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
+func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("reconciling flux-framework/fluence-controller for request")
+	pg := &schedv1alpha1.PodGroup{}
+
+	// Get the timestamp as soon as reconcile happens as a fallback below
+	timestamp := metav1.NewMicroTime(time.Now())
+	if err := r.Get(ctx, req.NamespacedName, pg); err != nil {
+
+		// Case 1: if we get here and it's not found, assume not created
+		if apierrs.IsNotFound(err) {
+			log.Info("PodGroup", "Status", fmt.Sprintf("Pod group %s is not found, deleted.", req.NamespacedName))
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, fmt.Sprintf("Unable to retrieve pod group %s", req.NamespacedName))
+		return ctrl.Result{}, err
+	}
+	log.Info("REFERENCES", "Reconciler", pg.ObjectMeta.OwnerReferences)
+
+	// Grab all statuses (and groups of them) we are interested in
+	schedulingOrPending := (pg.Status.Phase == schedv1alpha1.PodGroupScheduling || pg.Status.Phase == schedv1alpha1.PodGroupPending)
+	twoDaysOld := pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour
```

Comment:
I'm not sure - I refactored this to be more readable from here: https://github.com/kubernetes-sigs/scheduler-plugins/blob/2a7767e4b85b80ba86d967d860855b05899bb488/pkg/controllers/podgroup_controller.go#L88. Based on the comment, it seems like garbage collection would happen, but I can't find strong evidence of that - all I found was [this comment](https://github.com/kubernetes/kubernetes/commit/1457d204646f393add70d5385b7294f462b6f0ed#diff-5cd5a45cc2c4e646f07cf821ef2f48e03ca886f4f52c70a7c6623aedef2a430eR77-R80) that seems scoped to a unit test that suggests there is some GC at 48 hours. It could be there is an undocumented period, or that is specific to their testing setup.

If it's arbitrary we could definitely change it. Actually, I like the idea that a pod group might have this be customizable (via a label). We can consider that for the future.

Reply:
Sounds good!